### PR TITLE
SC quam with Qualibrate nodes

### PR DESCRIPTION
**Pull request by TheoLaudatQM**
SC calibration nodes with QuAM and Qualibrate.
Reviewing steps:
- Comprehensive readme with explanations about the folder structure and the installation guide
- Test the generation of quam and ensure that it is user friendly (make wiring, make_quam and modify_quam)
- Have a clean and working requirements.txt for easing the installation process --> wirer in the main of py-qua-tools + official qualibrate
- Figure out the structure of quam-libs: separate repo? In qualang-tools? Unversioned reference code? (required even if we only merge the generation of quam)

### CODE REVIEW

File: Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/flux_line.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+
+@quam_dataclass
+class FluxLine(SingleChannel):
+    """Example QuAM component for a flux line.
+
+    Args:
+        independent_offset (float): the flux bias for which the .
+        joint_offset (float): the flux bias for which the .
+        min_offset (float): the flux bias for which the .
```
Comment (nulinspiratie) [inline Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/flux_line.py:16]: Sentences are cut off

File: Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/transmon_pair.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+
+__all__ = ["TransmonPair"]
+
+
+@quam_dataclass
+class TransmonPair(QuamComponent):
+    id: Union[int, str]
+    qubit_control: Transmon = None
+    qubit_target: Transmon = None
```
Comment (nulinspiratie) [inline Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/transmon_pair.py:16]: We haven't reached concensus yet on whether to use qubit_control & qubit_target, or something else

Comment (nulinspiratie) [reply Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/transmon_pair.py:16]: It's fine to merge, just so you know this may become a breaking change in the future

File: Quantum-Control-Applications-QuAM/Superconducting/README.md [hunk 3]
```diff
@@ -0,0 +1,357 @@
+# N Flux-Tunable Transmon Qubits
+## Installation
+This folder contains an installable module called `quam_libs`, which provides a collection of tailored components for controlling flux-tunable qubits and experiment functionality. These components extend the functionality of QuAM, making it easier to design and execute calibration nodes.
+
+### Requirements
+To run the calibration nodes in this folder, you need to install `quam_libs`. First, ensure you have Python ≥ 3.8 installed on your system.
```
[the remark this answers is not in the document]
Comment (yomach) [reply Quantum-Control-Applications-QuAM/Superconducting/README.md:6]: In about 2 weeks we'll support 3.12.

Generally, I'm against writing the requirements in the Readme, you can see them in pypi and having to maintain it in multiple places will lead to errors.

[the remark this answers is not in the document]
Comment (yomach) [reply Quantum-Control-Applications-QuAM/Superconducting/README.md:6]: And in about 1-2 months we will stop supporting 3.8

[the remark this answers is not in the document]
Comment (KevinAVR) [reply Quantum-Control-Applications-QuAM/Superconducting/README.md:6]: yeah I thought about that, then let's not put it in the README :)

File: Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/transmon_pair.py (hunk 2 above)
Comment (MichalGQM) [reply Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/transmon_pair.py:16]: Let's decide on the name then? control and target looks okay to me

File: Quantum-Control-Applications-QuAM/Superconducting/README.md [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+   ```sh
+   sudo nano /etc/launchd.conf
+   ```
+3. Add the line:
+   ```sh
+   setenv QUAM_STATE_PATH "/path/to/configuration/quam_state"
+   ```
+4. Save the file and restart your system for changes to take effect.
+
+#### Windows
```
Comment (KevinAVR) [inline Quantum-Control-Applications-QuAM/Superconducting/README.md:54]: I tested this on Windows and works well

File: Quantum-Control-Applications-QuAM/Superconducting/README.md [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+1. All resonator I/Q channels are allocated to the first FEM/OPX+ for all qubits.
+2. All qubit I/Q channels are allocated to the first FEM/OPX+ for all qubits.
+3. All qubit flux channels are allocated to the first FEM/OPX+ for all qubits (if any).
+4. All tunable coupler channels are allocated to the first FEM/OPX+ for all qubits (if any).
+
+This extends over multiple LF-FEMs, OPX+ and Octaves when needed.
+
+## Folder structure
+
+The typical QUAM/QUalibrate folder structure is as follows:
```
Comment (KevinAVR) [inline Quantum-Control-Applications-QuAM/Superconducting/README.md:75]: this is clear :)

File: Quantum-Control-Applications-QuAM/Superconducting/README.md [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+* [modify_quam.py](./configuration/modify_quam.py): update the parameters of the state programmatically based on defaults values (previous calibration, chip manufacturer specification...).
+
+### quam_libs
+This folder contains all the utility functions necessary to create the wiring or build the QUAM, as well as QUA macros and data processing tools:
+* [components](./quam_libs/components): this is where the QUAM root and custom QUAM components are defined. A set of basic QUAM components are already present, but advanced user can easily modify them or create new ones.
+* [lib](./quam_libs/lib): contains several utility functions for saving, fitting and post-processing data.
+* [quam_builder](./quam_libs/quam_builder): contains the main functions called in [machine.py](./quam_libs/quam_builder/machine.py) and used to generate the wiring and build the QUAM structure from it and the QUAM components declared in the [components](./quam_libs/components) folder. It also contains the [pulses.py](./quam_libs/quam_builder/pulses.py) file where the default qubits pulses are defined.
+
```
Comment (KevinAVR) [inline Quantum-Control-Applications-QuAM/Superconducting/README.md:120]: Let's add the following comment for completion:

CosineDrag (with draft_coefficient = 0) and Square pulses are declared in pulses.py for the X and Y rotations, and by default operations such as x90, x180, -x90, y90, -y90, and y180 point to the CosineDrag pulse.

File: Quantum-Control-Applications-QuAM/Superconducting/README.md [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+* [modify_quam.py](./configuration/modify_quam.py): update the parameters of the state programmatically based on defaults values (previous calibration, chip manufacturer specification...).
+
+### quam_libs
+This folder contains all the utility functions necessary to create the wiring or build the QUAM, as well as QUA macros and data processing tools:
+* [components](./quam_libs/components): this is where the QUAM root and custom QUAM components are defined. A set of basic QUAM components are already present, but advanced user can easily modify them or create new ones.
+* [lib](./quam_libs/lib): contains several utility functions for saving, fitting and post-processing data.
+* [quam_builder](./quam_libs/quam_builder): contains the main functions called in [machine.py](./quam_libs/quam_builder/machine.py) and used to generate the wiring and build the QUAM structure from it and the QUAM components declared in the [components](./quam_libs/components) folder. It also contains the [pulses.py](./quam_libs/quam_builder/pulses.py) file where the default qubits pulses are defined.
```
Comment (KevinAVR) [inline Quantum-Control-Applications-QuAM/Superconducting/README.md:119]: In terms of wording for the paragraph I find it a little confusing that machine.py generates the wiring, I thought we generate the wiring when we called make_wiring_xxx.py in `configuration` -- @deanpoulos can it be rephrased different?

File: Quantum-Control-Applications-QuAM/Superconducting/README.md [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+
+## How to generate the QuAM
+Before starting to run experiments, it is necessary to build the Quantum Abstract Machine ([QuAM](https://github.com/qua-platform/quam)) for the desired 
+qubit chip architecture. More details about QUAM itself can be found in the [QuAM documentation](https://qua-platform.github.io/quam/).
+
+The process can be divided into three sections:
+1. First one needs to create the wiring specifying which quantum element (qubit, resonator, flux line...) is connected to 
+which OPX channel. 
+2. Then, the QuAM components can be derived from the resulting wiring.json file.
```
Comment (KevinAVR) [inline Quantum-Control-Applications-QuAM/Superconducting/README.md:128]: @deanpoulos Yonatan Cohen made a comment that because make_wiring_xxx.py is already declaring the quantum elements of the chip that the user will be working with, asked if we can take a pause and reconsider if a make_wiring_xxx.py would be replaced by make_chip_layout_xxx.py? even though the output of it is to generate wiring.json indeed. But the quam will be ruled to work with the qubits declared by make_wiring

File: Quantum-Control-Applications-QuAM/Superconducting/README.md [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+The process can be divided into three sections:
+1. First one needs to create the wiring specifying which quantum element (qubit, resonator, flux line...) is connected to 
+which OPX channel. 
+2. Then, the QuAM components can be derived from the resulting wiring.json file. 
+The attributes (methods or parameters) of these components can be customized at will. 
+3. Finally, a file called state.json, containing all the desired parameters that describe the full state of the system 
+will be created.
+
+### [1. Define the wiring](./configuration/make_wiring.py)
+The wiring is generated from a python script called [make_wiring.py](./configuration/make_wiring.py).
```
Comment (KevinAVR) [inline Quantum-Control-Applications-QuAM/Superconducting/README.md:134]: @TheoLaudatQM there is actually no make_wiring.py file, but there are three make_wiring for the different instruments set ups

File: Quantum-Control-Applications-QuAM/Superconducting/README.md [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+
+The hierarchy and structure of QUAM can be detailed as follows: 
+1. [quam_root.py](./quam_libs/components/quam_root.py) represents the highest level in terms of hierarchy.
+It contains the qubits and qubit pairs objects, as well as the wiring, network and Octaves. 
+Its methods are usually applied to all qubits or active qubits. 
+2. The definition of a QuAM transmon is defined in [transmon.py](./quam_libs/components/transmon.py). It contains the
+genral transmon attributes (T1, T2, f_01...) as well as the QuAM components composing the transmon (``xy``, ``resonator`` and 
+`z` in this case). Two-qubit gates can also be implemented by defining a specific qubit pair component as shown in 
+[transmon_pair.py](./quam_libs/components/transmon_pair.py).
+3. The sub-QuAM components are either defined from the base QUAM components directly, such as the qubit xy drive which is
```
Comment (KevinAVR) [inline Quantum-Control-Applications-QuAM/Superconducting/README.md:202]: replace sub-QuAM by `root QuAM componenets`, I believe that is the name given by @nulinspiratie  right?

File: Quantum-Control-Applications-QuAM/Superconducting/README.md [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+directly defined as an `IQChannel`, or from user-defined components such as 
+[readout_resonator](./quam_libs/components/readout_resonator.py) or [flux_line](./quam_libs/components/flux_line.py), 
+which allows the customization of their attributes.
+
+
+### [3. Generating the QuAM and state.json](./configuration/make_quam.py)
+Once the QuAM root is implemented with the subsequent components, the QuAM state can be generated automatically and each
```
Comment (KevinAVR) [inline Quantum-Control-Applications-QuAM/Superconducting/README.md:209]: I do not follow what subsequent components are being referred to here

File: Quantum-Control-Applications-QuAM/Superconducting/README.md [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+### [3. Generating the QuAM and state.json](./configuration/make_quam.py)
+Once the QuAM root is implemented with the subsequent components, the QuAM state can be generated automatically and each 
+parameter of the QUAM components is initialized to its arbitrary default value. 
+All of these parameters can be updated programmatically based on the specs from the chip manufacturer for instance and 
+the process is described in the next section.
+
+The script called [make_quam.py](./configuration/make_quam.py) takes care of generating the QUAM and filling the state 
+according to the wiring file. The Octaves connection parameters can also be edited here if relevant.
+
+Everything happens in the `build_quam` function, which programmatically adds all the Octaves, ports, transmons and pulses
```
Comment (KevinAVR) [inline Quantum-Control-Applications-QuAM/Superconducting/README.md:217]: Replace `Everything happens` -> `The QuAM generation happens` -- to be more accurate and explanatory

File: Quantum-Control-Applications-QuAM/Superconducting/README.md [hunk 13]
```diff
@@ -0,0 +1,357 @@
...
+
+# octave_settings = {"octave1": {"port": 11250} }  # externally configured: (11XXX where XXX are last three digits of oct ip)
+# octave_settings = {"oct1": {"ip": "192.168.88.250"} }  # "internally" configured: use the local ip address of the Octave
+octave_settings = {}
+
+# Make the QuAM object and save it
+quam = build_quam(machine, quam_state_path=path, octaves_settings=octave_settings)
+```
+
+Note that the set of default pulses can be edited in [pulses.py](./quam_libs/quam_builder/pulses.py).
```
Comment (KevinAVR) [inline Quantum-Control-Applications-QuAM/Superconducting/README.md:237]: add `Note that the set of default pulses, e. g. CosineDrag and Square, can be edited in ...

File: Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/flux_line.py [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+    """
+    independent_offset: float = 0.0
+    joint_offset: float = 0.0
+    min_offset: float = 0.0
+    arbitrary_offset: float = 0.0
```
[the remark this answers is not in the document]
Comment (KevinAVR) [reply Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/flux_line.py:21]: @nulinspiratie -- `quad_term` comes from the fitting to the qubit spec vs flux -- and `arbitrary_offset` is if you want to go experimentally to a given position and then you input it from qualibrate.

File: Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/tunable_coupler.py [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+    upsampling_mode: str = "pulse"
+
+    decouple_offset: float = 0.0
+    interaction_offset: float = 0.0
+
+    def __post_init__(self):
+        if isinstance(self.opx_output, LFFEMAnalogOutputPort):
+            self.opx_output.upsampling_mode = self.upsampling_mode
+            self.opx_output.output_mode = self.output_mode
+
```
[the remark this answers is not in the document]
Comment (KevinAVR) [reply Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/tunable_coupler.py:28]: @nulinspiratie seems reasonable to me that the coupler inherets properties from the nerighboor flux channels actually since its very likely the behaviors will replicate. Either they are 1e9 or 2e9 or mw or pulsed --

File: Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/readout_resonator.py [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+    f_12: float = None
+    confusion_matrix: list = None
+    
+    gef_centers : list = None
+    gef_confusion_matrix : list = None
+
+
+@quam_dataclass
+class ReadoutResonatorIQ(InOutIQChannel, ReadoutResonatorBase):
+    time_of_flight = 28  # smallest deviation from default (24ns) to work with Qualibrate
```
[the remark this answers is not in the document]
Comment (KevinAVR) [reply Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/readout_resonator.py:30]: I think so as well -- it is pending a test and then lets delete it

File: Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/readout_resonator.py [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+    # def get_output_power(self, operation, Z=50) -> float:
+    #     power = self.frequency_converter_up.power
+    #     amplitude = self.operations[operation].amplitude
+    #     x_mw = 10 ** (power / 10)
+    #     x_v = amplitude * np.sqrt(2 * Z * x_mw / 1000)
+    #     return 10 * np.log10(((x_v / np.sqrt(2)) ** 2 * 1000) / Z)
+
+@quam_dataclass
+class ReadoutResonatorMW(InOutMWChannel, ReadoutResonatorBase):
+    time_of_flight = 28
```
[the remark this answers is not in the document]
Comment (KevinAVR) [reply Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/readout_resonator.py:45]: same comment as above

File: Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/transmon.py [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+
+    f_01: float = None
+    f_12: float = None
+    anharmonicity: int = 150e6
+    freq_vs_flux_01_quad_term : float = 0.0
+    arbitrary_intermediate_frequency : float = 0.0
+
+    T1: int = 10_000
+    T2ramsey: int = 10_000
+    T2echo: int = 10_000
```
[the remark this answers is not in the document]
Comment (KevinAVR) [reply Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/transmon.py:48]: lets do NONE then, can you write down your proposed solution for @property?

File: Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/flux_line.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (nulinspiratie) [reply Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/flux_line.py:21]: Ok thanks, it would be helpfulto describe this in the docstring

File: Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/tunable_coupler.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (nulinspiratie) [reply Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/tunable_coupler.py:28]: @KevinAVR in this case it's not inheriting from the neighbouring flux channels. Instead, the coupler channel is taking these properties from its port. And this behaviour leads to issues.

File: Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/transmon_pair.py (hunk 2 above)
Comment (nulinspiratie) [reply Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/transmon_pair.py:16]: The root question is whether it makes sense in general for two-qubit gates to have a control and target. There are gates for which this doesn't make sense (e.g. symmetric gates such as CZ). I think it still makes sense to use `control` and `target`, but we need to reach alignment with the Arch team. Should be resolved soon, for now let's stick to this

File: Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/transmon.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (nulinspiratie) [reply Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/transmon.py:48]: Talked to Theo, I think the conclusion was to keep nanoseconds, but change type from int to float, right @TheoLaudatQM ?

[the remark this answers is not in the document]
Comment (nulinspiratie) [reply Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/transmon.py:48]: I would propose the following:
```
class Qubit(QuamComponent):
    ...
    T1: Optional[float] = None
    T2ramsey: Optional[float] = None
    T2echo: Optional[float] = None

    thermalization_factor: float = 5
    thermalization_time: int = "#./inferred_thermalization_time"

    @property
    def inferred_thermalization_time(self) -> int:
        return int(self.thermalization_factor * self.T2ramsey)
```

File: Quantum-Control-Applications-QuAM/Superconducting/quam_libs/quam_builder/wiring/create_analog_ports.py [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+    if channel.io_type == "output":
+        key = "frequency_converter_up"
+    elif channel.io_type == "input":
+        key = "frequency_converter_down"
+    else:
+        raise ValueError(f"Unknown IO type {channel.io_type}")
+
+    # todo: handle "octave" in OPX+ vs "oct" in new QOP?
+    reference = OCTAVES_BASE_JSON_PATH
+    reference += f"/octave{channel.con}"
```
Comment (TheoLaudatQM) [inline Quantum-Control-Applications-QuAM/Superconducting/quam_libs/quam_builder/wiring/create_analog_ports.py:20]: This will throw an error with QOP242 for which the Octave name is octx, to avoid this issue, why not specifying the octave_settings `octave_settings = {"oct1": {"ip": "192.168.88.250"} }` or `octave_settings = {"octave1": {"port": 11235} }` under make_wiring instead of make_quam? Then the `build_quam_wiring` function will be able to get the name from there
What do you guys think @deanpoulos @nulinspiratie ?

File: Quantum-Control-Applications-QuAM/Superconducting/README.md (hunk 6 above)
Comment (nulinspiratie) [reply Quantum-Control-Applications-QuAM/Superconducting/README.md:157]: I think we should remove `pulses.py` altogether, the drag cosine pulse is already added to quam

File: Quantum-Control-Applications-QuAM/Superconducting/README.md [hunk 20]
```diff
@@ -0,0 +1,357 @@
...
+### [2. The QuAM components](./quam_libs/components)
+Describe the structure of the QuAM components and how they can be customized, what to pay attention to...
+Also, how to create a new one and update the wiring and build_quam accordingly.
+
+The hierarchy and structure of QUAM can be detailed as follows: 
+1. [quam_root.py](./quam_libs/components/quam_root.py) represents the highest level in terms of hierarchy.
+It contains the qubits and qubit pairs objects, as well as the wiring, network and Octaves. 
+Its methods are usually applied to all qubits or active qubits. 
+2. The definition of a QuAM transmon is defined in [transmon.py](./quam_libs/components/transmon.py). It contains the
+genral transmon attributes (T1, T2, f_01...) as well as the QuAM components composing the transmon (``xy``, ``resonator`` and
```
Comment (nulinspiratie) [inline Quantum-Control-Applications-QuAM/Superconducting/README.md]: "general"

File: Quantum-Control-Applications-QuAM/Superconducting/README.md (hunk 10 above)
Comment (nulinspiratie) [reply Quantum-Control-Applications-QuAM/Superconducting/README.md:239]: No, the root QuAM component specifically refers to the top-level component inheriting from `QuamRoot`, but I agree that `sub-QuAM` can be confusing.
I would suggest to simply call this `QUAM components`

**Review by MichalGQM — APPROVED** [review]

File: Quantum-Control-Applications-QuAM/Superconducting/README.md (hunk 9 above)
Comment (TheoLaudatQM) [reply Quantum-Control-Applications-QuAM/Superconducting/README.md:171]: Good point, I just fixed it

File: Quantum-Control-Applications-QuAM/Superconducting/README.md (hunk 12 above)
Comment (TheoLaudatQM) [reply Quantum-Control-Applications-QuAM/Superconducting/README.md:257]: Done

File: Quantum-Control-Applications-QuAM/Superconducting/README.md [hunk 21]
```diff
@@ -0,0 +1,448 @@
...
+qualibrate start
+```
+3. Open http://localhost:8001/ on your browser:
+   ![browser window](./.img/qualibrate_1.png)
+4. Select the node you would like to run:
+   ![select node](./.img/qualibrate_2.png)
+5. Change the input parameters to your liking:
+   ![change parameters](./.img/qualibrate_3.png)
+6. Press "Run":
+   ![change parameters](./.img/qualibrate_4.png)
```
Comment (TheoLaudatQM) [inline Quantum-Control-Applications-QuAM/Superconducting/README.md:451]: I think that these screenshots are great and it would be awesome to have another one showing how to update the parameters at the end :)

File: Quantum-Control-Applications-QuAM/Superconducting/README.md [hunk 22]
```diff
@@ -0,0 +1,448 @@
...
+   json.dump(machine.generate_config(), f, indent=4)
+````
+
+Note that these parameters serve as a starting point before starting to calibrate the chip and their values will be
+updated at the end of each calibration node.
+
+## How to run Qualibrate nodes
+### Node structure
+> **_NOTE:_**  For the most detailed and up-to-date documentation on calibration nodes, visit the QUAlibrate [documentation](https://qua-platform.github.io/qualibrate/calibration_nodes/).
+>
```
Comment (TheoLaudatQM) [inline Quantum-Control-Applications-QuAM/Superconducting/README.md:391]: Instead I would rephrase it as `"For more details about Qualibrate and calibration nodes, visit the QUAlibrate [documentation](https://qua-platform.github.io/qualibrate/calibration_nodes/)."` to avoid leading the customers to think that the Github readmes are not up to date

File: Quantum-Control-Applications-QuAM/Superconducting/README.md (hunk 20 above)
Comment (TheoLaudatQM) [reply Quantum-Control-Applications-QuAM/Superconducting/README.md]: Fixed

File: Quantum-Control-Applications-QuAM/Superconducting/README.md (hunk 11 above)
Comment (TheoLaudatQM) [reply Quantum-Control-Applications-QuAM/Superconducting/README.md]: Good point

File: Quantum-Control-Applications-QuAM/Superconducting/README.md (hunk 7 above)
Comment (TheoLaudatQM) [reply Quantum-Control-Applications-QuAM/Superconducting/README.md:159]: In practice the wiring is indeed created when running make_wiring_xxx.py, but the function responsible for this is called `build_quam_wiring` and defined in machine.py.
@KevinAVR Do you mean renaming this file machine.py into something else?

File: Quantum-Control-Applications-QuAM/Superconducting/README.md (hunk 8 above)
Comment (TheoLaudatQM) [reply Quantum-Control-Applications-QuAM/Superconducting/README.md]: I understand Yonatan's comment, but I don't think that chip_layout is good, but we also declare the control hardware that is connected to the chip, so it is not only about the content of the chip.
What about `make_chip_connectivity_xxx.py`?
I'll create a poll to get people's opinions :)

File: Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/flux_line.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (TheoLaudatQM) [reply Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/flux_line.py:21]: I added it and removed quas_term because it is actually set at the qubit level

File: Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/flux_line.py [hunk 23]
```diff
@@ -0,0 +1,39 @@
+from quam.components import SingleChannel
+from quam.components.ports import LFFEMAnalogOutputPort
```
Comment (TheoLaudatQM) [inline Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/flux_line.py:2]: @nulinspiratie when toggling between OPX+ and LF fem, there is no need to change anything here right?
I can thus remove this line: `from quam.components.ports import LFFEMAnalogOutputPort`?

File: Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/flux_line.py (hunk 1 above)
Comment (TheoLaudatQM) [reply Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/flux_line.py:16]: Fixed

File: Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/readout_resonator.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (TheoLaudatQM) [reply Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/readout_resonator.py:30]: I just tested it and we can delete

File: Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/readout_resonator.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (TheoLaudatQM) [reply Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/readout_resonator.py:45]: removed as well

File: Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/transmon.py [hunk 24]
```diff
@@ -0,0 +1,145 @@
...
+        name = getattr(self, "name", self.__class__.__name__)
+        if not isinstance(self.f_01, (float, int)):
+            raise AttributeError(f"Error inferring anharmonicity for channel {name}: {self.f_01=} is not a number")
+        if not isinstance(self.f_12, (float, int)):
+            raise AttributeError(f"Error inferring anharmonicity for channel {name}: {self.f_12=} is not a number")
+        return self.f_12 - self.f_01
+
+    @property
+    def sigma(self, operation: Pulse):
+        return operation.length / self.sigma_time_factor
```
Comment (TheoLaudatQM) [inline Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/transmon.py:86]: Is this used anywhere?
I don't see its purpose and usage

File: Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/transmon.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (TheoLaudatQM) [reply Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/transmon.py:48]: Yes let's do this

File: Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/flux_line.py (hunk 23 above)
Comment (nulinspiratie) [reply Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/flux_line.py:2]: It doesn't seem to be used in the file, so I would say yes it's fine to remove

Comment (TheoLaudatQM) [reply Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/flux_line.py:2]: Yeah but what I mean is, if we now build quam with LF fem instead of OPX+, can we still use `SingleChannel` or do we need to change the component to be `FluxLine(LFFEMAnalogOutputPort`) instead?

Comment (nulinspiratie) [reply Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/flux_line.py:2]: No, the channels are separated from the ports, so no changes needed

Comment (nulinspiratie) [reply Quantum-Control-Applications-QuAM/Superconducting/quam_libs/components/flux_line.py:2]: It's a different matter for OPX + Octave versus MW FEM. Unfortunately there the changes are too significant between the two situations so it couldn't be combined into a single class

**Comment by KevinAVR** [comment]
I think if it is not in `main` and neither `iqcc-research` everything should go away and we can start with this basis
